core/src/avm2/verify.rs

```diff
@@ -561,279 +670,607 @@ fn optimize<'gc>(
-                        | Op::Decrement
-                        | Op::Negate => {
-                            previous_op = Some(op.clone());
-                            *op = Op::Nop;
-                            continue;
-                        }
-                        _ => {}
-                    },
-                    Op::CoerceI => match previous_op_some {
-                        Op::CoerceI | Op::PushByte { .. } | Op::PushShort { .. } => {
-                            previous_op = Some(op.clone());
+        if jump_targets.contains(&(i as i32)) {
+            stack.clear();
+            local_types = initial_local_types.clone();
+        }
+
+        match op {
+            Op::CoerceB => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Boolean)) {
+                    *op = Op::Nop;
+                }
+                stack.push_boolean();
+            }
+            Op::CoerceD => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Number)) {
+                    *op = Op::Nop;
+                }
+                stack.push_number();
+            }
+            Op::CoerceI => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Int))
+                    || matches!(stack_value, Some(ValueType::Uint))
+                {
+                    *op = Op::Nop;
+                }
+                stack.push_int();
+            }
+            Op::CoerceU => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Uint)) {
+                    *op = Op::Nop;
+                }
+                stack.push_uint();
+            }
+            Op::CoerceA => {
+                stack.pop();
+                stack.push_any();
+            }
+            Op::CoerceS => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Null)) {
+                    *op = Op::Nop;
+                }
+                stack.push_class_object(activation.avm2().classes().string);
+            }
+            Op::Equals | Op::LessEquals | Op::LessThan | Op::GreaterThan | Op::GreaterEquals => {
+                stack.pop();
+                stack.pop();
+                stack.push_boolean();
+            }
+            Op::Not => {
+                stack.pop();
+                stack.push_boolean();
+            }
+            Op::PushTrue | Op::PushFalse => {
+                stack.push_boolean();
+            }
+            Op::PushNull => {
+                stack.push_null();
+            }
+            Op::PushUndefined => {
+                stack.push_any();
+            }
+            Op::PushNaN => {
+                stack.push_number();
+            }
+            Op::PushByte { value } => {
+                if *value >= 0 {
+                    stack.push_uint();
```

Comment:
Fwiw, I'm not really sure if this matters. I think numeric types can almost freely swap with each other at runtime, so the difference between int and uint is mostly irrelevant for both the verifier and optimizer and we can't really freely optimize on them.

(Actually, this is super important in a completely different place - since a `Number`-typed value can be an `int`, we need to make sure that all three numeric types have identical disp_id->method binding, otherwise optimizing say `.toExponential()` can call a wrong method depending on runtime type.)

Reply:
This matters for `coerce_u`/`coerce_i`